File: Remove Nth Node From End of List.cpp

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
       int counter=0;
        int delet=0;
        int trace=0;
        ListNode* count=head;
        ListNode* current=head->next;
        ListNode* previous=head;

        while(count !=NULL) {
            count=count->next;
            counter++;
        }

        delet=(counter-n);
        int counter1=1;
        cout << delet;
        if(previous->next == nullptr){
            head = nullptr;
        } else if(delet == 0){
            head = head->next;
        }
        while(counter1 < delet) {
            previous = previous->next;
            current = current->next;
            counter1++ ;
        }
        
        if( delet + 1 == counter){
           
            previous->next = nullptr;
            
        } else if(current->next != nullptr){
           previous->next=current->next;
           
        }
        
        return head;
    } 
    
};

File: remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy=ListNode(0,head)
        right=head
        left=dummy
        i=0
        while right and i<n:
            right =right.next
            i=i+1
        while right:
            right=right.next
            left=left.next
        left.next=left.next.next
        return dummy.next